core/src/main/java/org/apache/iceberg/hadoop/JdbcTableDao.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hadoop;
+
+import java.sql.SQLException;
+import java.util.List;
+import org.apache.commons.dbutils.QueryRunner;
+import org.apache.commons.dbutils.handlers.BeanHandler;
+import org.apache.commons.dbutils.handlers.BeanListHandler;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class JdbcTableDao {
+  public static final String TABLES_TABLE_NAME = "iceberg_tables";
+  public static final String JDBC_CATALOG_TABLE_DDL =
+          "CREATE TABLE IF NOT EXISTS " + JdbcTableDao.TABLES_TABLE_NAME +
+                  "(catalogName VARCHAR(1255) NOT NULL," +
+                  "tableNamespace VARCHAR(1255) NOT NULL," +
+                  "tableName VARCHAR(1255) NOT NULL," +
+                  "metadataLocation VARCHAR(32768)," +
+                  "previousMetadataLocation VARCHAR(32768)," +
+                  "PRIMARY KEY (catalogName,tableNamespace,tableName)," +
+                  "FOREIGN KEY (catalogName, tableNamespace) REFERENCES " + JdbcNamespaceDao.NAMESPACES_TABLE_NAME +
```

Comment:
Is this `PRIMARY KEY` and `FOREIGN KEY` syntax portable? I'm wondering if we should omit them to have more generic SQL.

Of course, that would require some way to ensure that table creation is atomic because the primary key constraint wouldn't be enforced for the table. That would probably require a transaction to check whether a table exists and insert it to be portable. Some databases support `ON CONFLICT DO NOTHING` or `INSERT IGNORE` but it doesn't look like the syntax is portable.

It may be more portable to use `PRIMARY KEY`, but we will need to validate concurrent behavior.

Reply:
`PRIMARY KEY (catalogName,tableNamespace,tableName)` is implemented in most relational databases. 
but second one might not be. its nice to have PK to avoid duplicate data. another option might be adding it with alter statement it shuld be supported by many dbs 
`ALTER TABLE my_table ADD PRIMARY KEY (columnx);`

removed `PRIMARY KEY` and `FOREIGN KEY` lines for now.

Reply:
we could ignore if  'ALTER TABLE my_table ADD PRIMARY KEY (columnx);' fails. so for many dbs we can get PK created

Reply:
I think the problem is that we need that primary key to ensure that there is only one entry for a given table identifier. We don't want two concurrent creates to collide and create two different tables.

Reply:
using [sqlfiddle](http://sqlfiddle.com/) for testing all databases there are (mysql, MS sql, postgresql, oracle )
- compatible having `PRIMARY KEY` in the table DDL. 
- MS SQL Server not supporting create Table `IF NOT EXISTS`